File: src/routes/api/ws/logs.ts
import { createFileRoute } from '@tanstack/react-router'
import { defineHooks } from 'crossws'
import { dockerClient } from '#/lib/docker/client'

export type LogEntry = {
  containerName: string
  message: string
  stream: 'stdout' | 'stderr'
  timestamp: string
}

type LogSession = {
  processes: ReturnType<typeof Bun.spawn>[]
}

const sessions = new Map<string, LogSession>()

const decoder = new TextDecoder()

function processStream(
  stream: ReadableStream<Uint8Array>,
  containerName: string,
  streamType: 'stdout' | 'stderr',
  send: (data: string) => void,
): Promise<void> {
  const reader = stream.getReader()
  let buf = ''
  return (async () => {
    try {
      // eslint-disable-next-line
      while (true) {
        const { done, value } = await reader.read()
        if (done) break
        buf += decoder.decode(value, { stream: true })
        const parts = buf.split('\n')
        buf = parts.pop() ?? ''
        for (const line of parts) {
          if (!line) continue
          const spaceIdx = line.indexOf(' ')
          const timestamp = spaceIdx > -1 ? line.slice(0, spaceIdx) : ''
          const message = spaceIdx > -1 ? line.slice(spaceIdx + 1) : line
          const entry: LogEntry = {
            containerName,
            message,
            stream: streamType,
            timestamp,
          }
          send(JSON.stringify({ type: 'log', ...entry }))
        }
      }
      if (buf) {
        send(
          JSON.stringify({
            type: 'log',
            containerName,
            message: buf,
            stream: streamType,
            timestamp: '',
          }),
        )
      }
    } finally {
      reader.releaseLock()
    }
  })()
}

const hooks = defineHooks({
  async message(peer, message) {
    try {
      const msg = message.json<
        { type: 'init'; stackName: string } | { type: 'close' }
      >()

      if (msg.type === 'init') {
        const existing = sessions.get(peer.id)
        if (existing) {
          for (const p of existing.processes) p.kill()
          sessions.delete(peer.id)
        }

        try {
          const containers = await dockerClient.listContainers({
            filters: JSON.stringify({
              label: [`com.docker.compose.project=${msg.stackName}`],
            }),
          })

          if (containers.length === 0) {
            peer.send(JSON.stringify({ type: 'end' }))
            return
          }

          const processes: ReturnType<typeof Bun.spawn>[] = []
          sessions.set(peer.id, { processes })

          let finished = 0
          const total = containers.length * 2

          const checkDone = () => {
            finished++
            if (finished >= total) {
              peer.send(JSON.stringify({ type: 'end' }))
              sessions.delete(peer.id)
            }
          }

          for (const info of containers) {
            const containerName =
              info.Names[0]?.replace(/^\//, '') ?? info.Id.slice(0, 12)
            const proc = Bun.spawn(
              [
                'docker',
                'logs',
                '--follow',
                '--timestamps',
                '--tail',
                '1000',
                info.Id,
              ],
              { stdout: 'pipe', stderr: 'pipe' },
            )
            processes.push(proc)

            processStream(proc.stdout, containerName, 'stdout', (data) =>
              peer.send(data),
            ).finally(checkDone)

            processStream(proc.stderr, containerName, 'stderr', (data) =>
              peer.send(data),
            ).finally(checkDone)
          }
        } catch (err) {
          peer.send(JSON.stringify({ type: 'error', message: String(err) }))
        }
        return
      }

      // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
      if (msg.type === 'close') {
        const session = sessions.get(peer.id)
        if (session) {
          for (const p of session.processes) p.kill()
          sessions.delete(peer.id)
        }
        return
      }
    } catch {
      // not JSON
    }
  },

  close(peer) {
    const session = sessions.get(peer.id)
    if (session) {
      for (const p of session.processes) p.kill()
      sessions.delete(peer.id)
    }
  },
})

export const Route = createFileRoute('/api/ws/logs')({
  server: {
    handlers: {
      GET: async () => {
        return Object.assign(
          new Response('WebSocket upgrade is required.', { status: 426 }),
          { crossws: hooks },
        )
      },
    },
  },
})

File: internal/docker/logs.go
package docker

import (
	"bufio"
	"context"
	"io"
	"os/exec"
	"strings"
	"sync"
)

func (s *Stacks) StreamStackLogs(ctx context.Context, stackName string) (<-chan LogEntry, error) {
	containers, err := s.client.ListContainers(ctx, false, stackName)
	if err != nil {
		return nil, err
	}

	out := make(chan LogEntry)

	var wg sync.WaitGroup
	for _, c := range containers {
		containerName := c.ID
		if len(c.Names) > 0 {
			containerName = strings.TrimPrefix(c.Names[0], "/")
		} else if len(c.ID) > 12 {
			containerName = c.ID[:12]
		}
		wg.Add(1)
		go func(cName string, containerID string) {
			defer wg.Done()
			args := []string{}
			if s.hasConfigFile() {
				args = append(args, "--config", s.configDir)
			}
			args = append(args, "logs", "--follow", "--timestamps", "--tail", "1000", containerID)
			cmd := exec.CommandContext(ctx, "docker", args...)
			cmd.Env = getDockerEnv()
			stdout, err := cmd.StdoutPipe()
			if err != nil {
				return
			}
			stderr, err := cmd.StderrPipe()
			if err != nil {
				return
			}
			if err := cmd.Start(); err != nil {
				return
			}
			var streamWG sync.WaitGroup
			streamWG.Add(2)
			emit := func(r io.Reader, stream string) {
				defer streamWG.Done()
				sc := bufio.NewScanner(r)
				sc.Buffer(make([]byte, 64*1024), 1024*1024)
				for sc.Scan() {
					line := sc.Text()
					if line == "" {
						continue
					}
					entry := LogEntry{
						ContainerName: cName,
						Stream:        stream,
					}
					entry.Timestamp, entry.Message, _ = strings.Cut(line, " ")
					if entry.Message == "" {
						entry.Message = entry.Timestamp
						entry.Timestamp = ""
					}
					select {
					case out <- entry:
					case <-ctx.Done():
						return
					}
				}
				if err := sc.Err(); err != nil && ctx.Err() == nil {
					select {
					case out <- LogEntry{ContainerName: cName, Stream: stream, Message: "log stream error: " + err.Error()}:
					case <-ctx.Done():
					}
				}
			}
			go emit(stdout, "stdout")
			go emit(stderr, "stderr")
			streamWG.Wait()
			_ = cmd.Wait()
		}(containerName, c.ID)
	}

	go func() {
		wg.Wait()
		close(out)
	}()

	return out, nil
}
